[core] gcs AddEvent: TaskExecution support

sampan-s-nayak opened the pull request:
This is a series of PRs that connects each worker's EventAggregator to GCS, completing the OneEvent implementation for task events.
For more details, see: https://docs.google.com/document/d/1WjdlKprMqLqyPvmR1OGRQNhCD93SWrCXj21jZXXUcSk/edit?tab=t.0#heading=h.mmmgp2sapmts.

-------

This PR adds the logic to convert TaskExecutionEvent into TaskEvent for the GcsOneEventConverter. This further completes `GcsTaskManager::HandleAddEvent`.

Test:
- CI

Commit:
[core][1event/01] skeleton of gcs AddEvent grpc endpoint

Signed-off-by: Cuong Nguyen <can@anyscale.com>

Commit:
[core][1event/01] skeleton of gcs AddEvent grpc endpoint

Signed-off-by: Cuong Nguyen <can@anyscale.com>

Commit:
[core][onevent/02] implementation of gcs AddEvent endpoint

Signed-off-by: Cuong Nguyen <can@anyscale.com>

Commit:
[core][oneevent/02] implementation

Signed-off-by: Cuong Nguyen <can@anyscale.com>

Commit:
wip

Signed-off-by: Cuong Nguyen <can@anyscale.com>

Commit:
wip

Signed-off-by: Cuong Nguyen <can@anyscale.com>

Commit:
gcs AddEvent: TaskExecution support

Signed-off-by: Cuong Nguyen <can@anyscale.com>

Commit:
Merge remote-tracking branch 'origin/can-event02' into can-event03

Signed-off-by: sampan <sampan@anyscale.com>

Commit:
fix formatting after merge

Signed-off-by: sampan <sampan@anyscale.com>

sampan-s-nayak commented on src/ray/gcs/gcs_server/test/gcs_ray_event_converter_test.cc:
```diff
@@ -122,5 +122,46 @@ TEST_F(GcsRayEventConverterTest, TestConvertWithDroppedTaskAttempts) {
+  google::protobuf::Timestamp ts;
+  ts.set_seconds(42);
+  ts.set_nanos(123456789);
+  (*exec_event.mutable_task_state())[5] = ts;
```
Ill try it out

sampan-s-nayak commented on src/ray/gcs/gcs_server/gcs_ray_event_converter.cc:
```diff
@@ -81,5 +85,23 @@ void GcsRayEventConverter::ConvertToTaskEvents(rpc::events::TaskDefinitionEvent
+  task_state_update->set_node_id(event.node_id());
+  task_state_update->set_worker_id(event.worker_id());
+  task_state_update->set_worker_pid(event.worker_pid());
+  *task_state_update->mutable_error_info() = std::move(event.ray_error_info());
```
interesting, let me verify the behaviour of std::move().

Commit:
address comments

Signed-off-by: sampan <sampan@anyscale.com>

sampan-s-nayak committed:
Merge branch 'can-event02' into can-event03

Signed-off-by: Sampan S Nayak <sampansnayak2@gmail.com>

sampan-s-nayak committed:
Merge branch 'can-event02' into can-event03

Commit:
Merge remote-tracking branch 'upstream/can-event02' into can-event03

Signed-off-by: sampan <sampan@anyscale.com>

sampan-s-nayak committed:
Merge branch 'can-event02' into can-event03

Commit:
Merge branch 'can-event02' into can-event03

Signed-off-by: sampan <sampan@anyscale.com>

MengjinYan commented on src/ray/gcs/gcs_server/gcs_ray_event_converter.cc:
```diff
@@ -133,5 +137,23 @@ void GcsRayEventConverter::ConvertToTaskEvents(rpc::events::TaskDefinitionEvent
+  task_state_update->mutable_error_info()->Swap(event.mutable_ray_error_info());
+
+  for (const auto &[state, timestamp] : event.task_state()) {
+    int64_t ns = timestamp.seconds() * 1000000000LL + timestamp.nanos();
```
Probably better to add a util function in `grpc_util.h` similar to the following for the conversion:
https://github.com/ray-project/ray/blob/bbbe9ef08a896d6512f750e8b6aaf489b0a77dd5/src/ray/common/grpc_util.h#L235-L244

sampan-s-nayak committed:
Merge branch 'can-event02' into can-event03

Commit:
address comment

Signed-off-by: sampan <sampan@anyscale.com>

Commit:
fix import

Signed-off-by: sampan <sampan@anyscale.com>

Commit:
fix lint

Signed-off-by: sampan <sampan@anyscale.com>

MengjinYan commented on src/ray/common/grpc_util.h:
```diff
@@ -243,4 +243,9 @@ inline google::protobuf::Timestamp AbslTimeNanosToProtoTimestamp(int64_t nanos)
   return timestamp;
 }
 
+inline int64_t ProtoTimestampToAbslTimeNanos(
```
```suggestion
// Conver a protobuf timestamp to an epoch time in nanoseconds
// Ref: https://protobuf.dev/reference/php/api-docs/Google/Protobuf/Timestamp.html
inline int64_t ProtoTimestampToAbslTimeNanos(
```

Commit:
Merge remote-tracking branch 'upstream/can-event02' into can-event03

Signed-off-by: sampan <sampan@anyscale.com>

Commit:
address comments

Signed-off-by: sampan <sampan@anyscale.com>

sampan-s-nayak committed:
[core] gcs AddEvent: ActorDefinition support (#55531)

This is a series of PRs that connects each worker's EventAggregator to
GCS, completing the OneEvent implementation for task events.
For more details, see:
https://docs.google.com/document/d/1WjdlKprMqLqyPvmR1OGRQNhCD93SWrCXj21jZXXUcSk/edit?tab=t.0#heading=h.mmmgp2sapmts.

-------

This PR adds the logic to convert
ActorTaskDefinitionEvent+ActorTaskExecutionEvent into TaskEvent for the
GcsOneEventConverter. This further completes
`GcsTaskManager::HandleAddEvent`.

Test:
- CI

---------

Signed-off-by: Cuong Nguyen <can@anyscale.com>
Signed-off-by: sampan <sampan@anyscale.com>
Co-authored-by: Cuong Nguyen <can@anyscale.com>
Co-authored-by: sampan <sampan@anyscale.com>
